# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

In [108]:
import numpy as np 
import pandas as pd 


In [109]:
df1 = pd.read_csv('e_commerce.csv')
df1

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

## 评估数据

In [110]:
df1.sample(10)

InvoiceNo StockCode                         Description  Quantity  \
297001    562933     23158         SET OF 5 LUCKY CAT MAGNETS          1   
360603    568239     22423            REGENCY CAKESTAND 3 TIER         1   
492361    578109     22959              WRAP CHRISTMAS VILLAGE        25   
303455    563546     23005    TRAVEL CARD WALLET I LOVE LONDON        48   
232576    557324     22617          BAKING SET SPACEBOY DESIGN         1   
82938     543276     22767         TRIPLE PHOTO FRAME CORNICE         16   
246588    558707     22244          3 HOOK HANGER MAGIC GARDEN         1   
501288    578827     21822   GLITTER CHRISTMAS TREE WITH BELLS         1   
228885    556964     23245          SET OF 3 REGENCY CAKE TINS         4   
149254    549283     22488  NATURAL SLATE RECTANGLE CHALKBOARD        12   

             InvoiceDate  UnitPrice  CustomerID         Country  
297001   8/10/2011 16:51       4.13         NaN  United Kingdom  
360603   9/26/2011 12:02      12.75     18205.0  United Kingdom  
492361   11/23/2011 9:22       0.42     14020.0  United Kingdom  
303455   8/17/2011 12:24       0.42     16729.0  United Kingdom  
232576    6/20/2011 9:41      10.79         NaN  United Kingdom  
82938     2/6/2011 13:57       8.50     16525.0  United Kingdom  
246588    7/1/2011 12:58       1.95     12748.0  United Kingdom  
501288  11/25/2011 14:25       1.63         NaN  United Kingdom  
228885    6/16/2011 9:46       4.95     16145.0  United Kingdom  
149254    4/7/2011 13:32       1.65     13005.0  United Kingdom

### 评估数据干净度

In [111]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


需要
1、处理缺失值
2、更改数据类型InvoiceDate应该是日期、CustomerID应该是字符串
## 评估缺失数据
在了解了```'Description'```存在缺失值后，根据条件提取出缺失观察值

In [112]:
df1[df1['Description'].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

有1454行的交易数据缺失```''Description''```的值
从输出结果来看，这些缺失```''Description''```的交易数据，```'UnitPrice'```都为0。
为了验证猜想，我们增加筛选条件，看是否存在```''Description''```变量缺失且```'UnitPrice'```不为0的数据

In [113]:
df1[(df1['Description'].isnull())&('UnitPrice' != 0 )]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

筛选出来的结构数量为0，说明缺失```''Description''```的值，同时也不具备有效的```'UnitPrice'```值。
前者代表商品名称，后者是单价，这两者缺失的数据显然也没有有效意义，认为可以后续被删除。

In [114]:
df1[df1['CustomerID'].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

```'CustomerID'```仅仅代表客户编号，不是分析商品的必要变量。并且从输出结果来看，有些CustomerID缺失的交易数据仍然有效，因此保留此变量为空的观察值。
## 评估重复数据
其实根据该数据集，任何变量的重复在实际购买行为中都有意义，无需评估重复数据

In [115]:
df1[df1.duplicated(subset=["InvoiceNo",'StockCode'])]

InvoiceNo StockCode                      Description  Quantity  \
125       536381     71270                  PHOTO CLIP LINE         3   
498       536409    90199C  5 STRAND GLASS NECKLACE CRYSTAL         1   
502       536409     85116  BLACK CANDELABRA T-LIGHT HOLDER         5   
517       536409     21866      UNION JACK FLAG LUGGAGE TAG         1   
525       536409    90199C  5 STRAND GLASS NECKLACE CRYSTAL         2   
...          ...       ...                              ...       ...   
541692    581538     22992           REVOLVER WOODEN RULER          1   
541697    581538     21194        PINK  HONEYCOMB PAPER FAN         1   
541698    581538    35004B      SET OF 3 BLACK FLYING DUCKS         1   
541699    581538     22694                     WICKER STAR          1   
541701    581538     23343     JUMBO BAG VINTAGE CHRISTMAS          1   

            InvoiceDate  UnitPrice  CustomerID         Country  
125      12/1/2010 9:41       1.25     15311.0  United Kingdom  
498     12/1/2010 11:45       6.35     17908.0  United Kingdom  
502     12/1/2010 11:45       2.10     17908.0  United Kingdom  
517     12/1/2010 11:45       1.25     17908.0  United Kingdom  
525     12/1/2010 11:45       6.35     17908.0  United Kingdom  
...                 ...        ...         ...             ...  
541692  12/9/2011 11:34       1.95     14446.0  United Kingdom  
541697  12/9/2011 11:34       0.65     14446.0  United Kingdom  
541698  12/9/2011 11:34       5.45     14446.0  United Kingdom  
541699  12/9/2011 11:34       2.10     14446.0  United Kingdom  
541701  12/9/2011 11:34       2.08     14446.0  United Kingdom  

[10684 rows x 8 columns]

## 评估不一致数据

不一致数据可能存在于```'Country'```变量中，我们要查看是否存在不同字符串指代同一国家的情况

In [116]:
df1['Country'].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


可以发现 ```'USA'``` 和 ```'United States'```  都表示美国，```'UK'``` 和 ```'U.K.'```还有
```'United Kingdom'``` 都是英国，应该保持一致。后续处理应该只保留一个指代值

## 评估无效或错误数据
可以通过Dataframe的```describe```方法，对数值统计信息进行快速了解

In [117]:
df1.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从数据结构看，```'Quantity'```和```'UnitPrice'```存在负数，我们把负数提取出来评估其意义

In [118]:
df1[df1['Quantity']<0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

可见，这些发票编号都是C开头的，根据要求，证明这些交易被取消，猜测取消的订单产品数量都是负数。
我们来尝试验证：

In [119]:
df1[(df1['Quantity']<0)&(df1['InvoiceNo'].str[0]!='C')]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

证明猜想错误，存在开头不是C的行```'Quantity	'```值也是负数
继续观察，发现```'UnitPrice'```都是0，猜测，不以C开头的```'InvoiceNo'```行，```'Quantity	'```值也是负数代表单价为0

In [120]:
df1[(df1['Quantity']<0)&(df1['InvoiceNo'].str[0]!='C')&(df1['UnitPrice']!=0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

发现猜想正确，证明```'Quantity'```为负数的商品都可以删除，后续数据清理的时候要删除。

接下来筛选```'UnitPrice'```为负数的观察值

In [121]:
df1[df1['UnitPrice']<0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

```df1['UnitPrice']<0```代表坏账调整，也可以删

## 清理数据
要做的事项有：
1. 改数据类型```'InvoiceDate'```应该是日期、```'CustomerID'```应该是字符串
2. 删除缺失```''Description''```的值
3.  ```'USA'``` 统一为 ```'United States'```  都表示美国，
4. ```'UK'``` 和 ```'U.K.'```统一为```'United Kingdom'``` 都是英国
5. ```'Quantity'```为负数的商品都可以删除
6. ```df1['UnitPrice']<0```的观察值擅长

In [122]:
clean_data = df1.copy()
clean_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [123]:
# 转换时间类型
clean_data['InvoiceDate'] = pd.to_datetime(clean_data['InvoiceDate'])
clean_data['InvoiceDate']

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

```'CustomerID'```转换类型为字符串

In [124]:
clean_data['CustomerID'] = clean_data['CustomerID'].astype(str)
clean_data['CustomerID'] 

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

In [125]:
# 为了删除结尾的.0，用到Python Pandas Series.str.slice()用法
clean_data['CustomerID'] = clean_data['CustomerID'].str.slice(0,-2)
clean_data['CustomerID']

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

删除缺失```''Description''```的值

In [126]:
clean_data.dropna(subset=['Description'],inplace=True)

In [127]:
clean_data['Description'].isnull().sum()

0

In [128]:
clean_data['Country']=clean_data['Country'].replace({'USA':'United States'})

In [129]:
len(clean_data[clean_data['Country']=='USA'])

0

把```'UK'``` 和 ```'U.K.'```统一为```'United Kingdom'``` 

In [130]:
clean_data['Country']=clean_data['Country'].replace({'UK':'United Kingdom','U.K.':'United Kingdom'})

In [131]:
print(len(clean_data[clean_data['Country']=='UK']))
print(len(clean_data[clean_data['Country']=='U.K.']))

0
0


5. ```'Quantity'```为负数的商品都可以删除
6. ```df1['UnitPrice']<0```的观察值删除

In [132]:
clean_data=clean_data[clean_data['Quantity']>=0]

In [135]:
clean_data[clean_data['Quantity']<0]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [136]:
clean_data=clean_data[clean_data['UnitPrice']>=0]

In [137]:
clean_data[clean_data['UnitPrice']<0]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [140]:
clean_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-12-09 12:50:00       0.85      12680          France  
541905 2011-12-09 12:50:00       2.10      12680          France  
541906 2011-12-09 12:50:00       4.15      12680          France  
541907 2011-12-09 12:50:00       4.15      12680          France  
541908 2011-12-09 12:50:00       4.95      12680          France  

[530691 rows x 8 columns]

## 保存清理后的数据

In [139]:
clean_data.to_csv('e_commerce_cleaned.csv',index=False)